runtime/src/cost_model.rs

```diff
@@ -100,9 +102,13 @@ impl CostModel {
         tx_cost
     }
 
-    pub fn upsert_instruction_cost(&mut self, program_key: &Pubkey, cost: u64) {
+    pub fn upsert_instruction_cost(
+        &mut self,
+        program_key: &Pubkey,
+        program_timing: &ProgramTiming,
```

Comment:
I find it unintuitive to squirl away the cost in a metrics tracking type, which typically has no bearing on logic. i think it would make more sense to pass `cost` and `count` in here and bubble up a `Result<(), CostCalculationError>` to somewhere more appropriate to be handling timing/metrics

Reply:
I feel bubbling up calculation error to up level for handling adds "soft coupling", it would be a stronger design choice if there is a central error handling component, in my opinion.

Reply:
> Mmm, I thought this interface change is better. The data flow is: `cost_update_service` receives `program_timing`, then calls this `upsert_instruction_cost(...)` to pass received data in order to update `cost` in `execute_cost_table`. This change lets `execute_cost_table` to define what `cost` should be, so `cost_update_service` can just do dispatching service, it really shouldn't have the concept of `cost`. Instead `execute_cost_table`'s upsert function neatly encapsulates what `cost` is and how it is calculated.

The problem is that AFAIK this is the only place in the codebase where we're overloading an ancillary `*Timings` for logical duties.  It's unexpected and will be easy to misread for the uninitiated.  If I see a method that's named so as to imply it's doing something with a cost, but has no `cost`-sounding parameters, now I have to go digging through code to figure out what's going on rather than engage intuition.  Further I think it sets a bad precedent as we really don't want this kind of mixed logical/ancillary data usage.

> I feel bubbling up calculation error to up level for handling adds "soft coupling", it would be a stronger design choice if there is a central error handling component, in my opinion.

it seems to me that overloading the `ProgramTimings` is a much tighter coupling, with far more potential for negative maintenance burden than returning an error in a case where quire frankly, an error has occurred.